### PR TITLE
Parse record io with custom decoder.

### PR DESCRIPTION
**Pull request by jeschkies**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,99 +1,284 @@
 extern crate bytes;
-#[macro_use]
 extern crate futures;
 extern crate hyper;
 extern crate tokio_core;
 
 mod mesos {
```
Comment (boxdot) [inline src/lib.rs]: A good practice is to create another file called `mesos.rs`, and just write here: `mod mesos`. Inside the file you are automatically inside the module and do not need indentation.

File: src/lib.rs [hunk 2]
```diff
@@ -1,99 +1,284 @@
...
-        pub body: hyper::Body,
+    pub trait Decoder<T> {
+        fn decode(&mut self, buf: &mut BytesMut) -> Option<T>;
     }
 
-    impl RecordIoConnection {
-        fn next_line(&mut self) -> Option<String> {
+    /// Decodes lines of body stream.
+    pub struct LineDecoder;
+    impl Decoder<String> for LineDecoder {
```
Comment (boxdot) [inline src/lib.rs]: Are you sure you want to allocate Strings for each line?

File: src/lib.rs [hunk 3]
```diff
@@ -1,99 +1,284 @@
...
-                let line = self.buf.split_to(i);
-                self.buf.split_to(1);
+            if let Some(i) = buf.iter().position(|&b| b == b'\n') {
+                let line = buf.split_to(i);
+                buf.split_to(1);
 
                 match str::from_utf8(&line) {
                     Ok(s) => return Some(s.to_string()),
-                    Err(e) => {
+                    Err(_) => {
```
Comment (boxdot) [inline src/lib.rs]: use failure::Error, the return type would be then `Result<Option<String>, failure::Error>`

File: src/lib.rs [hunk 4]
```diff
@@ -1,99 +1,284 @@
...
+        state: RecordIoDecoderState,
+    }
+    impl RecordIoDecoder {
+        pub fn new() -> Self {
+            Self {
+                state: RecordIoDecoderState::TrimWhitespaces,
+            }
+        }
+
+        fn is_whitespace(&mut self, b: &u8) -> bool {
```
Comment (boxdot) [inline src/lib.rs]: Does not need to be mut.

File: src/lib.rs [hunk 5]
```diff
@@ -1,99 +1,284 @@
...
+                        return RecordIoDecoderState::ReadLength;
+                    }
+                };
+            }
+            RecordIoDecoderState::ReadLength
+        }
+
+        pub fn decode_record(
+            &mut self,
+            length: u64,
```
Comment (boxdot) [inline src/lib.rs]: usize if what is used for length normally

**Review by boxdot — APPROVED** [review]

File: src/lib.rs (hunk 5 above)
Comment (jeschkies) [reply src/lib.rs]: The RecordIO specs state it as `u64`. The question is when to do the conversion.

**Comment by boxdot** [comment]
Looks good!

**Pull request merged**